common/cuda_hip/matrix/fbcsr_kernels.hpp.inc

```diff
@@ -73,3 +73,86 @@ __global__ __launch_bounds__(default_block_size) void transpose_blocks(
 
 
 }  // namespace kernel
+
+
+template <typename ValueType, typename IndexType>
+void from_matrix_data(
+    std::shared_ptr<const DefaultExecutor> exec,
+    const Array<matrix_data_entry<ValueType, IndexType>>& data, int block_size,
+    Array<IndexType>& row_ptr_array, Array<IndexType>& col_idx_array,
+    Array<ValueType>& value_array)
+{
+    using entry = matrix_data_entry<ValueType, IndexType>;
+    using device_entry = device_type<entry>;
+    const auto nnz = data.get_num_elems();
+    const auto bs = block_size;
+    auto row_ptrs = row_ptr_array.get_data();
+    auto num_rows = row_ptr_array.get_num_elems() - 1;
+    if (nnz == 0) {
+        components::fill_array(exec, row_ptrs, num_rows, IndexType{});
```

Comment:
Shouldn't this be `num_rows + 1`?
```suggestion
        components::fill_array(exec, row_ptrs, num_rows + 1, IndexType{});
```

Reply:
Yep, due to all initializations going through Reference, we aren't actually testing the device-side read functions right now. Working on fixing that.

Reply:
Actually, I'll add tests and potential fixes in #904